File: robot_position_cmd/src/motoman_robot.cpp
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit_msgs/GetPositionIK.h>
#include <std_msgs/Int8.h>
#include <geometry_msgs/Pose.h>
#include <motoman_msgs/WriteSingleIO.h>

void move_take_img_pose_cb(const std_msgs::Int8ConstPtr msg);
void target_point_cb(const std_msgs::Int8ConstPtr msg);
void place_piece_cb(const std_msgs::Int8ConstPtr msg);
void move_start_cb(const std_msgs::Int8ConstPtr msg);
int getIK(moveit_msgs::GetPositionIK & server, ros::ServiceClient & client, 
    moveit::planning_interface::MoveGroupInterface & group, 
    geometry_msgs::Pose & target_pose, std::vector<double> & joint_group_positions);
void move_robot_joints(moveit::planning_interface::MoveGroupInterface & move_group, 
    std::vector<double> & target_joints);
void define_task_positions();
void move_robot_to_pose(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group, geometry_msgs::Pose & pose);
void move_to_take_img_pose(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group);
void move_to_target_point(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group);
void move_to_place_piece(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group);
void move_to_start(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group);
void set_vaccum(int status1, int status2);

// publishers
ros::Publisher move_ref_pose_pub;
ros::Publisher pick_up_complete_pub;
ros::Publisher place_complete_pub;
ros::Publisher task_complete_pub;


static const std::string PLANNING_GROUP = "arm_on_rail";
static const std::string MOVE_GROUP = "mh6";

geometry_msgs::PoseStamped start_pose_stmp;
std::vector<double> joint_move_group_positions;
sensor_msgs::JointState joint_state;
std::vector<double> home_position, aux_position;

bool move_to_reference_pose = false, move_to_pick_up = false, 
    move_to_place = false, move_start_pos = false;

// task positions
geometry_msgs::Pose reference_pose, pick_up_pose, place_pose;

// servers
moveit_msgs::GetPositionIK ik_server;
motoman_msgs::WriteSingleIO io_server;
// clients
ros::ServiceClient ik_client;
ros::ServiceClient io_client;


int main(int argc, char * argv[]){
    ros::init(argc, argv, "motoman_robot");
    ros::NodeHandle nh;
    ros::AsyncSpinner spinner(1);
    spinner.start();

    // planning group    
    moveit::planning_interface::MoveGroupInterface group(PLANNING_GROUP);
    group.setPlannerId("RRTConnectkConfigDefault");
    moveit::core::RobotStatePtr current_state_group = group.getCurrentState();
    const robot_state::JointModelGroup* joint_model_group = group.getCurrentState()->getJointModelGroup(PLANNING_GROUP);
    std::vector<double> joint_group_positions;
    current_state_group->copyJointGroupPositions(joint_model_group, joint_group_positions);
    start_pose_stmp = group.getCurrentPose(group.getEndEffectorLink());


    // move group
    moveit::planning_interface::MoveGroupInterface move_group(MOVE_GROUP);
    move_group.setPlannerId("RRTConnectkConfigDefault");
    moveit::core::RobotStatePtr current_state_move_group = move_group.getCurrentState();
    const robot_state::JointModelGroup* joint_model_move_group = move_group.getCurrentState()->getJointModelGroup(MOVE_GROUP);
    
    current_state_group->copyJointGroupPositions(joint_model_move_group, joint_move_group_positions);

    define_task_positions();

    // subscribers
    ros::Subscriber move_ref_pose_sub = nh.subscribe("/move_reference_pose", 1, move_take_img_pose_cb);
    ros::Subscriber target_point_sub = nh.subscribe("/reference_pose_reached", 1, target_point_cb);
    ros::Subscriber place_piece_sub = nh.subscribe("/pick_up_completed", 1, place_piece_cb);
    ros::Subscriber start_pose_cmd_sub = nh.subscribe("/place_completed", 1, move_start_cb);

    // publishers
    move_ref_pose_pub = nh.advertise<std_msgs::Int8>("/reference_pose_reached", 1);
    pick_up_complete_pub = nh.advertise<std_msgs::Int8>("/pick_up_completed", 1);
    place_complete_pub = nh.advertise<std_msgs::Int8>("/place_completed", 1);
    task_complete_pub = nh.advertise<std_msgs::Int8>("/task_completed", 1);

    
    // clients
    ik_client = nh.serviceClient<moveit_msgs::GetPositionIK>("compute_ik");
    io_client = nh.serviceClient<motoman_msgs::WriteSingleIO>("write_single_io");

    while(ros::ok()){
        if(move_to_reference_pose){
            move_to_reference_pose = false;
            move_to_take_img_pose(group, move_group);
            ros::Duration(0.2).sleep();
            move_to_target_point(group, move_group);
            ros::Duration(0.2).sleep();
            move_to_place_piece(group, move_group);
            ros::Duration(0.2).sleep();
            move_to_start(group, move_group);
        }
/*
        if(move_to_reference_pose){
            move_to_reference_pose = false;
            move_to_take_img_pose(group, move_group);
        }else if(move_to_pick_up){
            move_to_pick_up = false;
            move_to_target_point(group, move_group);
        }else if(move_to_place){
            move_to_place = false;
            move_to_place_piece(group, move_group);
        }else if(move_start_pos){
            move_start_pos = false;
            move_to_start(group, move_group);
        }
*/
        ros::spinOnce();
    }

    // ros::spin();
    ros::shutdown();
    return 0;
}


void move_to_take_img_pose(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group){

    ROS_INFO("Moving the robot to reference pose");
    move_robot_joints(move_group, aux_position);

    move_robot_to_pose(group, move_group, reference_pose);

    std_msgs::Int8 reference_reached;
    reference_reached.data = 1;
    ros::Duration(0.2).sleep();
    // move_ref_pose_pub.publish(reference_reached);
}

void move_take_img_pose_cb(const std_msgs::Int8ConstPtr msg){
    move_to_reference_pose = true;    
}


void move_to_target_point(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group){

    ROS_INFO("Moving the robot to pick up pose");
    geometry_msgs::Pose pick_up_sup_pose = pick_up_pose;
    pick_up_sup_pose.position.z += 0.1;
    move_robot_to_pose(group, move_group, pick_up_sup_pose);

    move_robot_to_pose(group, move_group, pick_up_pose);
    
    // vaccum ON
    ros::Duration(1.0).sleep();
    set_vaccum(0, 1);
    ros::Duration(1.0).sleep();
    
    move_robot_to_pose(group, move_group, pick_up_sup_pose);
    ROS_INFO("Moving back the robot to reference pose");
    move_robot_to_pose(group, move_group, reference_pose);

    move_robot_joints(move_group, aux_position);
    move_robot_joints(move_group, home_position);

    std_msgs::Int8 pick_up_complete;
    pick_up_complete.data = 1;
    ros::Duration(0.2).sleep();
    // pick_up_complete_pub.publish(pick_up_complete);
}

void target_point_cb(const std_msgs::Int8ConstPtr msg){
    move_to_pick_up = true;
}


void move_to_place_piece(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group){
    ROS_INFO("Moving the robot to place pose");
    geometry_msgs::Pose place_sup_pose = place_pose;
    place_sup_pose.position.z += 0.3;
    move_robot_to_pose(group, move_group, place_sup_pose);

    place_sup_pose.position.z -= 0.2;
    move_robot_to_pose(group, move_group, place_sup_pose);

    move_robot_to_pose(group, move_group, place_pose);
    // vaccum OFF
    ros::Duration(1.0).sleep();
    set_vaccum(1, 0);
    ros::Duration(1.0).sleep();
    
    ROS_INFO("Moving back the robot to reference pose");
    move_robot_to_pose(group, move_group, place_sup_pose);

    place_sup_pose.position.z += 0.2;
    move_robot_to_pose(group, move_group, place_sup_pose);

    std_msgs::Int8 place_piece_complete;
    place_piece_complete.data = 1;
    ros::Duration(0.2).sleep();
    // place_complete_pub.publish(place_piece_complete);
}

void place_piece_cb(const std_msgs::Int8ConstPtr msg){
    move_to_place = true;
}


void move_to_start(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group){
    ROS_INFO("Moving the robot to start pose");
    move_robot_joints(move_group, home_position);

    std_msgs::Int8 task_complete;
    task_complete.data = 1;
    ros::Duration(0.2).sleep();
    // task_complete_pub.publish(task_complete);
}


void move_start_cb(const std_msgs::Int8ConstPtr msg){
    move_start_pos = true;
}


int getIK(moveit_msgs::GetPositionIK & server, ros::ServiceClient & client, 
    moveit::planning_interface::MoveGroupInterface & group, 
    geometry_msgs::Pose & target_pose, std::vector<double> & joint_group_positions){
    
    // calculate IK
    server.request.ik_request.group_name = group.getName();
    server.request.ik_request.ik_link_name = group.getEndEffectorLink();
    
    // set desired target pose stamped
    geometry_msgs::PoseStamped target_pose_stmp = start_pose_stmp;
    target_pose_stmp.pose = target_pose;
    server.request.ik_request.pose_stamped = target_pose_stmp;

    server.request.ik_request.timeout = ros::Duration(10.0);
    server.request.ik_request.attempts = 0;

    client.call(server);
    if(server.response.error_code.val == 1){
        joint_state = server.response.solution.joint_state;
        for(int i = 0; i < joint_state.position.size(); i++){
            joint_group_positions[i] = joint_state.position[i];
        }
        return 1;
    }else{
        ROS_INFO("IK solution not found");
        return 0;
    }
}


void move_robot_to_pose(moveit::planning_interface::MoveGroupInterface & group, 
    moveit::planning_interface::MoveGroupInterface & move_group, geometry_msgs::Pose & pose){
    if(getIK(ik_server, ik_client, group, pose, joint_move_group_positions)){
        joint_move_group_positions[7] = 0.0;
        move_robot_joints(move_group, joint_move_group_positions);
    }
}


void move_robot_joints(moveit::planning_interface::MoveGroupInterface & move_group, 
    std::vector<double> & target_joints){
    move_group.setJointValueTarget(target_joints);
    moveit::planning_interface::MoveGroupInterface::Plan my_plan;
    bool success = (move_group.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
    if(success) move_group.move();
}


void set_vaccum(int status1, int status2){
	
	io_server.request.address = 10032;
	io_server.request.value = status1;
	if(io_client.call(io_server)){
        // std::cout << "Reset 10032" << std::endl;
    }else{
        ROS_ERROR("Failed to call service write_single_io");
	}

    io_server.request.address = 10033;
    io_server.request.value = status2;
    if(io_client.call(io_server)){
        // std::cout << "Set 10033" << std::endl;
    }else{
        ROS_ERROR("Failed to call service write_single_io");
    }	
}


void define_task_positions() {
    reference_pose.position.x = 1.7356;
    reference_pose.position.y = -0.9611;
    reference_pose.position.z = 1.3875;
    reference_pose.orientation.x = 0.706583;
    reference_pose.orientation.y = -0.707629;
    reference_pose.orientation.z = 0.000545;
    reference_pose.orientation.w = 0.000545;

    pick_up_pose.position.x = 1.8024;
    pick_up_pose.position.y = -0.9818;
    pick_up_pose.position.z = 0.3405;
    pick_up_pose.orientation.x = 0.706636;
    pick_up_pose.orientation.y = -0.707576;
    pick_up_pose.orientation.z = 0.000473;
    pick_up_pose.orientation.w = 0.000485;

    place_pose.position.x = 4.5289;
    place_pose.position.y = -0.1248;
    place_pose.position.z = 0.7300;
    place_pose.orientation.x = 0.999999;
    place_pose.orientation.y = 0.000783;
    place_pose.orientation.z = 0.000807;
    place_pose.orientation.w = 0.0;

    for(int i = 0; i < 8; i++){
        home_position.push_back(0.0);
        aux_position.push_back(0.0);
    }
    aux_position[0] = -1.0;
}

File: robot_position_cmd/scripts/motoman_robot.py
#!/usr/bin/env python
import sys
import copy
import rospy
import moveit_commander
from geometry_msgs.msg import Pose
from std_msgs.msg import Int8

from moveit_msgs.srv import GetPositionIK
from moveit_msgs.msg import PositionIKRequest

from motoman_msgs.srv import WriteSingleIO

class robot:
    def __init__(self):
        self.robot = moveit_commander.RobotCommander()
        self.group = moveit_commander.MoveGroupCommander('arm_on_rail')
        self.group.set_planner_id('RRTConnectkConfigDefault')
        self.group_move = moveit_commander.MoveGroupCommander('mh6')
        self.group_move.set_planner_id('RRTConnectkConfigDefault')
        self.start_pose = self.group.get_current_pose('arm_link_tool0')
        self.start_pose_modificable = self.group.get_current_pose('arm_link_tool0')
        self.home_position = [0, 0, 0, 0, 0, 0, 0, 0]
        self.aux_position = [-1.0, 0, 0, 0, 0, 0, 0, 0]

        # ========================== robot poses for task ==========================
        # reference pose is where photo is taken
        self.reference_pose = Pose()
        self.reference_pose.position.x = 1.7356
        self.reference_pose.position.y = -0.9611
        self.reference_pose.position.z = 1.3975
        self.reference_pose.orientation.x = 0.706583
        self.reference_pose.orientation.y = -0.707629
        self.reference_pose.orientation.z = 0.000545
        self.reference_pose.orientation.w = 0.000545
        # -------------------------------------------------------------------------
        # pick_up pose is where robot takes the piece
        self.pick_up_pose = Pose()
        self.pick_up_pose.position.x = 1.8024
        self.pick_up_pose.position.y = -0.9818
        self.pick_up_pose.position.z = 0.3595
        self.pick_up_pose.orientation.x = 0.706636
        self.pick_up_pose.orientation.y = -0.707576
        self.pick_up_pose.orientation.z = 0.000473
        self.pick_up_pose.orientation.w = 0.000485        
        # -------------------------------------------------------------------------
        # place pose is where robot places the piece
        self.place_pose = Pose()
        self.place_pose.position.x = 4.5289
        self.place_pose.position.y = -0.1248
        self.place_pose.position.z = 0.7400
        self.place_pose.orientation.x = 0.999999
        self.place_pose.orientation.y = 0.000783
        self.place_pose.orientation.z = 0.000807
        self.place_pose.orientation.w = 0.0
        # =========================================================================

        # subscribers
        self.move_ref_pose_sub = rospy.Subscriber('/move_reference_pose', Int8, self.move_take_img_pose_cb, queue_size = 1)
        # center point is target center in image
        self.target_point_sub = rospy.Subscriber('/reference_pose_reached', Int8, self.target_point_cb, queue_size = 1)
        self.place_piece_sub = rospy.Subscriber('/pick_up_completed', Int8, self.place_piece_cb, queue_size = 1)
        self.start_pose_cmd_sub = rospy.Subscriber('/place_completed', Int8, self.move_start_cb, queue_size = 1)

        # publishers
        self.move_ref_pose_pub = rospy.Publisher('/reference_pose_reached', Int8, queue_size = 1)
        self.pick_up_complete_pub = rospy.Publisher('/pick_up_completed', Int8, queue_size = 1)
        self.place_complete_pub = rospy.Publisher('/place_completed', Int8, queue_size = 1)
        self.task_complete_pub = rospy.Publisher('/task_completed', Int8, queue_size = 1)


    def move_start_cb(self, data):
        rospy.loginfo('Moving the robot to start pose')
        print('self.start_pose.pose', self.start_pose.pose)
        self.move_robot_joint(self.home_position)
        task_complete = Int8()
        task_complete.data = 1
        self.task_complete_pub.publish(task_complete)


    def place_piece_cb(self, data):
        rospy.loginfo('Moving the robot to place pose')
        place_sup_pose = Pose(position = self.place_pose.position, orientation = self.place_pose.orientation)
        z_aux = self.place_pose.position.z
        place_sup_pose.position.z += 0.30
        self.move_robot(place_sup_pose)

        place_sup_pose.position.z -= 0.20
        self.move_robot(place_sup_pose)

        place_sup_pose.position.z -= 0.10        
        self.move_robot(self.place_pose)
        # vaccum OFF
        rospy.sleep(2)
        self.set_vaccum(1, 0)
        rospy.sleep(2)

        self.move_robot(place_sup_pose)

        place_sup_pose.position.z += 0.20
        self.move_robot(place_sup_pose)

        self.place_pose.position.z = z_aux

        place_piece_complete = Int8()
        place_piece_complete.data = 1
        self.place_complete_pub.publish(place_piece_complete)


    def target_point_cb(self, data):
        rospy.loginfo('Moving the robot to pick up pose')
        pick_up_sup_pose = Pose(position = self.pick_up_pose.position, orientation = self.pick_up_pose.orientation)
        z_aux = self.pick_up_pose.position.z
        pick_up_sup_pose.position.z += 0.10
        self.move_robot(pick_up_sup_pose)

        self.pick_up_pose.position.z = z_aux

        self.move_robot(self.pick_up_pose)
        print('=========================== pick_up_pose ==========================')
        print(self.pick_up_pose)
        print('===================================================================')
        # vaccum ON
        rospy.sleep(2)
        self.set_vaccum(0, 1)
        rospy.sleep(2)

        self.move_robot(pick_up_sup_pose)

        rospy.loginfo('Moving back the robot to reference pose')
        self.move_robot(self.reference_pose)

        self.move_robot_joint(self.aux_position)
        self.move_robot_joint(self.home_position)

        self.pick_up_pose.position.z = z_aux

        pick_up_complete = Int8()
        pick_up_complete.data = 1
        self.pick_up_complete_pub.publish(pick_up_complete)


    def move_take_img_pose_cb(self, data):
        rospy.loginfo('Moving the robot to reference pose')
        self.move_robot_joint(self.aux_position)

        self.move_robot(self.reference_pose)
        
        reference_reached = Int8()
        reference_reached.data = 1
        self.move_ref_pose_pub.publish(reference_reached)


    def move_robot(self, target_pose):
        ik = self.get_ik(target_pose)
        if(ik.error_code.val == 1):
            joints = []
            print(len(ik.solution.joint_state.position))
            for i in range(len(ik.solution.joint_state.position)):
                joints.append(ik.solution.joint_state.position[i])
            self.move_robot_joint(joints)
        else:
            rospy.loginfo('IK not found')


    def move_robot_joint(self, target_joints):
        self.group_move.set_joint_value_target(target_joints)
        self.group_move.go()
        

    def get_ik(self, pose):
        rospy.loginfo('Waiting for IK service')
        rospy.wait_for_service('compute_ik')
        
        ik_request = PositionIKRequest()
        ik_request.group_name = self.group.get_name()
        ik_request.robot_state = self.robot.get_current_state()
        ik_request.ik_link_name = self.group.get_end_effector_link()
        print('ik_request.ik_link_name: ', ik_request.ik_link_name)
        
        pose_stmp = self.start_pose_modificable
        pose_stmp.pose = pose
        ik_request.pose_stamped = pose_stmp
        ik_request.timeout.secs = 10
        ik_request.attempts = 0
        try:
            ik_service = rospy.ServiceProxy('compute_ik', GetPositionIK)
            resp = ik_service(ik_request)
            print(resp.solution.joint_state)
            return resp
        except rospy.ServiceException, e:
            rospy.logerr('Error with IK service')


    def set_vaccum(self, status1, status2):
        rospy.loginfo('Waiting for IO service')
        # rospy.wait_for_service('write_single_io')
        try:
            io_service = rospy.ServiceProxy('write_single_io', WriteSingleIO)
            io_service(10032, status1)
            io_service(10033, status2)
        except rospy.ServiceException, e:
            rospy.logerr('Error with IO service')


    def print_pose(self, pose):
        print("============ Start pose ============")
        print("Position: ", pose.pose.position.x, 
            pose.pose.position.y, 
            pose.pose.position.z)
        print("Orientation: ", pose.pose.orientation.x, 
            pose.pose.orientation.y, 
            pose.pose.orientation.z, 
            pose.pose.orientation.w)
        print("====================================")
        

def main():
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('motoman_robot', anonymous = True)
    rospy.loginfo("Starting node")
    my_robot = robot()
    print("================ Ready ================")
    try:
        rospy.spin()
    except KeyboardInterrupt:
        pass
    moveit_commander.roscpp_shutdown()


if __name__ == '__main__':
    main()
